// @pendle/core-v2@5.5.0/typechain-types/factories/I1inchAggregationRouterV6__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import { Provider } from "@ethersproject/providers";
import type {
  I1inchAggregationRouterV6,
  I1inchAggregationRouterV6Interface,
} from "../I1inchAggregationRouterV6";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "executor",
        type: "address",
      },
      {
        components: [
          {
            internalType: "contract IERC20",
            name: "srcToken",
            type: "address",
          },
          {
            internalType: "contract IERC20",
            name: "dstToken",
            type: "address",
          },
          {
            internalType: "address payable",
            name: "srcReceiver",
            type: "address",
          },
          {
            internalType: "address payable",
            name: "dstReceiver",
            type: "address",
          },
          {
            internalType: "uint256",
            name: "amount",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "minReturnAmount",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "flags",
            type: "uint256",
          },
        ],
        internalType: "struct I1inchAggregationRouterV6.SwapDescription",
        name: "desc",
        type: "tuple",
      },
      {
        internalType: "bytes",
        name: "data",
        type: "bytes",
      },
    ],
    name: "swap",
    outputs: [
      {
        internalType: "uint256",
        name: "returnAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "spentAmount",
        type: "uint256",
      },
    ],
    stateMutability: "payable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "I1inchCommonType.Address256",
        name: "to",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "token",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "minReturn",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "dex",
        type: "uint256",
      },
    ],
    name: "unoswapTo",
    outputs: [
      {
        internalType: "uint256",
        name: "returnAmount",
        type: "uint256",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "I1inchCommonType.Address256",
        name: "to",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "token",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "minReturn",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "dex",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "dex2",
        type: "uint256",
      },
    ],
    name: "unoswapTo2",
    outputs: [
      {
        internalType: "uint256",
        name: "returnAmount",
        type: "uint256",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "I1inchCommonType.Address256",
        name: "to",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "token",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "minReturn",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "dex",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "dex2",
        type: "uint256",
      },
      {
        internalType: "I1inchCommonType.Address256",
        name: "dex3",
        type: "uint256",
      },
    ],
    name: "unoswapTo3",
    outputs: [
      {
        internalType: "uint256",
        name: "returnAmount",
        type: "uint256",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class I1inchAggregationRouterV6__factory {
  static readonly abi = _abi;
  static createInterface(): I1inchAggregationRouterV6Interface {
    return new utils.Interface(_abi) as I1inchAggregationRouterV6Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): I1inchAggregationRouterV6 {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as I1inchAggregationRouterV6;
  }
}
