// @pendle/core-v2@5.5.0/typechain-types/factories/IKyberScalingHelper__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import { Provider } from "@ethersproject/providers";
import type {
  IKyberScalingHelper,
  IKyberScalingHelperInterface,
} from "../IKyberScalingHelper";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes",
        name: "inputData",
        type: "bytes",
      },
      {
        internalType: "uint256",
        name: "newAmount",
        type: "uint256",
      },
    ],
    name: "getScaledInputData",
    outputs: [
      {
        internalType: "bool",
        name: "isSuccess",
        type: "bool",
      },
      {
        internalType: "bytes",
        name: "data",
        type: "bytes",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

export class IKyberScalingHelper__factory {
  static readonly abi = _abi;
  static createInterface(): IKyberScalingHelperInterface {
    return new utils.Interface(_abi) as IKyberScalingHelperInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IKyberScalingHelper {
    return new Contract(address, _abi, signerOrProvider) as IKyberScalingHelper;
  }
}
