// @pendle/core-v2@5.5.0/typechain-types/factories/IPActionVePendleStatic__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import { Provider } from "@ethersproject/providers";
import type {
  IPActionVePendleStatic,
  IPActionVePendleStaticInterface,
} from "../IPActionVePendleStatic";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "user",
        type: "address",
      },
      {
        internalType: "uint128",
        name: "additionalAmountToLock",
        type: "uint128",
      },
      {
        internalType: "uint128",
        name: "newExpiry",
        type: "uint128",
      },
    ],
    name: "increaseLockPositionStatic",
    outputs: [
      {
        internalType: "uint128",
        name: "newVeBalance",
        type: "uint128",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

export class IPActionVePendleStatic__factory {
  static readonly abi = _abi;
  static createInterface(): IPActionVePendleStaticInterface {
    return new utils.Interface(_abi) as IPActionVePendleStaticInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IPActionVePendleStatic {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as IPActionVePendleStatic;
  }
}
