// @pendle/core-v2@5.5.0/typechain-types/factories/IPActionStorageStatic__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import { Provider } from "@ethersproject/providers";
import type {
  IPActionStorageStatic,
  IPActionStorageStaticInterface,
} from "../IPActionStorageStatic";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyUsed",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPtUsed",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpOut",
        type: "uint256",
      },
    ],
    name: "AddLiquidityDualSyAndPt",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "tokenIn",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenUsed",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPtUsed",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "AddLiquidityDualTokenAndPt",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPtIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpOut",
        type: "uint256",
      },
    ],
    name: "AddLiquiditySinglePt",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpOut",
        type: "uint256",
      },
    ],
    name: "AddLiquiditySingleSy",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyMintPy",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netYtOut",
        type: "uint256",
      },
    ],
    name: "AddLiquiditySingleSyKeepYt",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "AddLiquiditySingleToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netYtOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyMintPy",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "AddLiquiditySingleTokenKeepYt",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpIn",
        type: "uint256",
      },
      {
        components: [
          {
            internalType: "uint256",
            name: "netPtFromRemove",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromRemove",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netPtRedeem",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromRedeem",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "totalSyOut",
            type: "uint256",
          },
        ],
        indexed: false,
        internalType: "struct ExitPostExpReturnParams",
        name: "params",
        type: "tuple",
      },
    ],
    name: "ExitPostExpToSy",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "totalTokenOut",
        type: "uint256",
      },
      {
        components: [
          {
            internalType: "uint256",
            name: "netPtFromRemove",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromRemove",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netPtRedeem",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromRedeem",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "totalSyOut",
            type: "uint256",
          },
        ],
        indexed: false,
        internalType: "struct ExitPostExpReturnParams",
        name: "params",
        type: "tuple",
      },
    ],
    name: "ExitPostExpToToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpIn",
        type: "uint256",
      },
      {
        components: [
          {
            internalType: "uint256",
            name: "netPtFromRemove",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromRemove",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netPyRedeem",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromRedeem",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netPtSwap",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netYtSwap",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromSwap",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFee",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "totalSyOut",
            type: "uint256",
          },
        ],
        indexed: false,
        internalType: "struct ExitPreExpReturnParams",
        name: "params",
        type: "tuple",
      },
    ],
    name: "ExitPreExpToSy",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "totalTokenOut",
        type: "uint256",
      },
      {
        components: [
          {
            internalType: "uint256",
            name: "netPtFromRemove",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromRemove",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netPyRedeem",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromRedeem",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netPtSwap",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netYtSwap",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFromSwap",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "netSyFee",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "totalSyOut",
            type: "uint256",
          },
        ],
        indexed: false,
        internalType: "struct ExitPreExpReturnParams",
        name: "params",
        type: "tuple",
      },
    ],
    name: "ExitPreExpToToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "YT",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPyOut",
        type: "uint256",
      },
    ],
    name: "MintPyFromSy",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "tokenIn",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "YT",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPyOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "MintPyFromToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "tokenIn",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "SY",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyOut",
        type: "uint256",
      },
    ],
    name: "MintSyFromToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "previousOwner",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "OwnershipTransferred",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "YT",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPyIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyOut",
        type: "uint256",
      },
    ],
    name: "RedeemPyToSy",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "tokenOut",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "YT",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPyIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "RedeemPyToToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "tokenOut",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "SY",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyIn",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenOut",
        type: "uint256",
      },
    ],
    name: "RedeemSyToToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpToRemove",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPtOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyOut",
        type: "uint256",
      },
    ],
    name: "RemoveLiquidityDualSyAndPt",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "tokenOut",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpToRemove",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPtOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "RemoveLiquidityDualTokenAndPt",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpToRemove",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netPtOut",
        type: "uint256",
      },
    ],
    name: "RemoveLiquiditySinglePt",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpToRemove",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyOut",
        type: "uint256",
      },
    ],
    name: "RemoveLiquiditySingleSy",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netLpToRemove",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netTokenOut",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "RemoveLiquiditySingleToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "bytes4",
        name: "selector",
        type: "bytes4",
      },
      {
        indexed: true,
        internalType: "address",
        name: "facet",
        type: "address",
      },
    ],
    name: "SelectorToFacetSet",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "netPtToAccount",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "netSyToAccount",
        type: "int256",
      },
    ],
    name: "SwapPtAndSy",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "netPtToAccount",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "netTokenToAccount",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "SwapPtAndToken",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "netYtToAccount",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "netSyToAccount",
        type: "int256",
      },
    ],
    name: "SwapYtAndSy",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "caller",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "market",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "token",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "receiver",
        type: "address",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "netYtToAccount",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "netTokenToAccount",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "netSyInterm",
        type: "uint256",
      },
    ],
    name: "SwapYtAndToken",
    type: "event",
  },
  {
    inputs: [],
    name: "claimOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "getDefaultApproxParams",
    outputs: [
      {
        components: [
          {
            internalType: "uint256",
            name: "guessMin",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "guessMax",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "guessOffchain",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "maxIteration",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "eps",
            type: "uint256",
          },
        ],
        internalType: "struct ApproxParams",
        name: "",
        type: "tuple",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "getOwnerAndPendingOwner",
    outputs: [
      {
        internalType: "address",
        name: "_owner",
        type: "address",
      },
      {
        internalType: "address",
        name: "_pendingOwner",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        components: [
          {
            internalType: "uint256",
            name: "guessMin",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "guessMax",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "guessOffchain",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "maxIteration",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "eps",
            type: "uint256",
          },
        ],
        internalType: "struct ApproxParams",
        name: "params",
        type: "tuple",
      },
    ],
    name: "setDefaultApproxParams",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
      {
        internalType: "bool",
        name: "direct",
        type: "bool",
      },
      {
        internalType: "bool",
        name: "renounce",
        type: "bool",
      },
    ],
    name: "transferOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class IPActionStorageStatic__factory {
  static readonly abi = _abi;
  static createInterface(): IPActionStorageStaticInterface {
    return new utils.Interface(_abi) as IPActionStorageStaticInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IPActionStorageStatic {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as IPActionStorageStatic;
  }
}
