// @pendle/core-v2@5.5.0/typechain-types/factories/IPActionStorageEvents__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import { Provider } from "@ethersproject/providers";
import type {
  IPActionStorageEvents,
  IPActionStorageEventsInterface,
} from "../IPActionStorageEvents";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "previousOwner",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "OwnershipTransferred",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "bytes4",
        name: "selector",
        type: "bytes4",
      },
      {
        indexed: true,
        internalType: "address",
        name: "facet",
        type: "address",
      },
    ],
    name: "SelectorToFacetSet",
    type: "event",
  },
];

export class IPActionStorageEvents__factory {
  static readonly abi = _abi;
  static createInterface(): IPActionStorageEventsInterface {
    return new utils.Interface(_abi) as IPActionStorageEventsInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IPActionStorageEvents {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as IPActionStorageEvents;
  }
}
