// @pendle/core-v2@5.5.0/typechain-types/factories/IPPausingInterface__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import { Provider } from "@ethersproject/providers";
import type {
  IPPausingInterface,
  IPPausingInterfaceInterface,
} from "../IPPausingInterface";

const _abi = [
  {
    inputs: [],
    name: "pause",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "paused",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "unpause",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class IPPausingInterface__factory {
  static readonly abi = _abi;
  static createInterface(): IPPausingInterfaceInterface {
    return new utils.Interface(_abi) as IPPausingInterfaceInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IPPausingInterface {
    return new Contract(address, _abi, signerOrProvider) as IPPausingInterface;
  }
}
