// @pendle/core-v2@5.5.0/typechain-types/IKyberScalingHelper.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import { FunctionFragment, Result } from "@ethersproject/abi";
import { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export interface IKyberScalingHelperInterface extends utils.Interface {
  contractName: "IKyberScalingHelper";
  functions: {
    "getScaledInputData(bytes,uint256)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "getScaledInputData",
    values: [BytesLike, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "getScaledInputData",
    data: BytesLike
  ): Result;

  events: {};
}

export interface IKyberScalingHelper extends BaseContract {
  contractName: "IKyberScalingHelper";
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IKyberScalingHelperInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    getScaledInputData(
      inputData: BytesLike,
      newAmount: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[boolean, string] & { isSuccess: boolean; data: string }>;
  };

  getScaledInputData(
    inputData: BytesLike,
    newAmount: BigNumberish,
    overrides?: CallOverrides
  ): Promise<[boolean, string] & { isSuccess: boolean; data: string }>;

  callStatic: {
    getScaledInputData(
      inputData: BytesLike,
      newAmount: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[boolean, string] & { isSuccess: boolean; data: string }>;
  };

  filters: {};

  estimateGas: {
    getScaledInputData(
      inputData: BytesLike,
      newAmount: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    getScaledInputData(
      inputData: BytesLike,
      newAmount: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
