// @pendle/core-v2@5.5.0/typechain-types/IPActionVePendleStatic.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import { FunctionFragment, Result } from "@ethersproject/abi";
import { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export interface IPActionVePendleStaticInterface extends utils.Interface {
  contractName: "IPActionVePendleStatic";
  functions: {
    "increaseLockPositionStatic(address,uint128,uint128)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "increaseLockPositionStatic",
    values: [string, BigNumberish, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "increaseLockPositionStatic",
    data: BytesLike
  ): Result;

  events: {};
}

export interface IPActionVePendleStatic extends BaseContract {
  contractName: "IPActionVePendleStatic";
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IPActionVePendleStaticInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    increaseLockPositionStatic(
      user: string,
      additionalAmountToLock: BigNumberish,
      newExpiry: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { newVeBalance: BigNumber }>;
  };

  increaseLockPositionStatic(
    user: string,
    additionalAmountToLock: BigNumberish,
    newExpiry: BigNumberish,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  callStatic: {
    increaseLockPositionStatic(
      user: string,
      additionalAmountToLock: BigNumberish,
      newExpiry: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  filters: {};

  estimateGas: {
    increaseLockPositionStatic(
      user: string,
      additionalAmountToLock: BigNumberish,
      newExpiry: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    increaseLockPositionStatic(
      user: string,
      additionalAmountToLock: BigNumberish,
      newExpiry: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
