// @pendle/core-v2@5.5.0/typechain-types/IPMultiTokenMerkleDistributor.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import { FunctionFragment, Result, EventFragment } from "@ethersproject/abi";
import { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export interface IPMultiTokenMerkleDistributorInterface
  extends utils.Interface {
  contractName: "IPMultiTokenMerkleDistributor";
  functions: {
    "claim(address,address[],uint256[],bytes32[][])": FunctionFragment;
    "claimVerified(address,address[])": FunctionFragment;
    "claimed(address,address)": FunctionFragment;
    "merkleRoot()": FunctionFragment;
    "verified(address,address)": FunctionFragment;
    "verify(address,address[],uint256[],bytes32[][])": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "claim",
    values: [string, string[], BigNumberish[], BytesLike[][]]
  ): string;
  encodeFunctionData(
    functionFragment: "claimVerified",
    values: [string, string[]]
  ): string;
  encodeFunctionData(
    functionFragment: "claimed",
    values: [string, string]
  ): string;
  encodeFunctionData(
    functionFragment: "merkleRoot",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "verified",
    values: [string, string]
  ): string;
  encodeFunctionData(
    functionFragment: "verify",
    values: [string, string[], BigNumberish[], BytesLike[][]]
  ): string;

  decodeFunctionResult(functionFragment: "claim", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "claimVerified",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "claimed", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "merkleRoot", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "verified", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "verify", data: BytesLike): Result;

  events: {
    "Claimed(address,address,address,uint256)": EventFragment;
    "SetDefaultTokenList(address[])": EventFragment;
    "SetMerkleRoot(bytes32)": EventFragment;
    "Verified(address,address,uint256)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "Claimed"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "SetDefaultTokenList"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "SetMerkleRoot"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "Verified"): EventFragment;
}

export type ClaimedEvent = TypedEvent<
  [string, string, string, BigNumber],
  { token: string; user: string; receiver: string; amount: BigNumber }
>;

export type ClaimedEventFilter = TypedEventFilter<ClaimedEvent>;

export type SetDefaultTokenListEvent = TypedEvent<
  [string[]],
  { defaultTokenList: string[] }
>;

export type SetDefaultTokenListEventFilter =
  TypedEventFilter<SetDefaultTokenListEvent>;

export type SetMerkleRootEvent = TypedEvent<[string], { merkleRoot: string }>;

export type SetMerkleRootEventFilter = TypedEventFilter<SetMerkleRootEvent>;

export type VerifiedEvent = TypedEvent<
  [string, string, BigNumber],
  { token: string; user: string; amountClaimable: BigNumber }
>;

export type VerifiedEventFilter = TypedEventFilter<VerifiedEvent>;

export interface IPMultiTokenMerkleDistributor extends BaseContract {
  contractName: "IPMultiTokenMerkleDistributor";
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IPMultiTokenMerkleDistributorInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    claim(
      receiver: string,
      tokens: string[],
      totalAccrueds: BigNumberish[],
      proofs: BytesLike[][],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    claimVerified(
      receiver: string,
      tokens: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    claimed(
      token: string,
      user: string,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    merkleRoot(overrides?: CallOverrides): Promise<[string]>;

    verified(
      token: string,
      user: string,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    verify(
      user: string,
      tokens: string[],
      totalAccrueds: BigNumberish[],
      proofs: BytesLike[][],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;
  };

  claim(
    receiver: string,
    tokens: string[],
    totalAccrueds: BigNumberish[],
    proofs: BytesLike[][],
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  claimVerified(
    receiver: string,
    tokens: string[],
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  claimed(
    token: string,
    user: string,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  merkleRoot(overrides?: CallOverrides): Promise<string>;

  verified(
    token: string,
    user: string,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  verify(
    user: string,
    tokens: string[],
    totalAccrueds: BigNumberish[],
    proofs: BytesLike[][],
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    claim(
      receiver: string,
      tokens: string[],
      totalAccrueds: BigNumberish[],
      proofs: BytesLike[][],
      overrides?: CallOverrides
    ): Promise<BigNumber[]>;

    claimVerified(
      receiver: string,
      tokens: string[],
      overrides?: CallOverrides
    ): Promise<BigNumber[]>;

    claimed(
      token: string,
      user: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    merkleRoot(overrides?: CallOverrides): Promise<string>;

    verified(
      token: string,
      user: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    verify(
      user: string,
      tokens: string[],
      totalAccrueds: BigNumberish[],
      proofs: BytesLike[][],
      overrides?: CallOverrides
    ): Promise<BigNumber[]>;
  };

  filters: {
    "Claimed(address,address,address,uint256)"(
      token?: string | null,
      user?: string | null,
      receiver?: string | null,
      amount?: null
    ): ClaimedEventFilter;
    Claimed(
      token?: string | null,
      user?: string | null,
      receiver?: string | null,
      amount?: null
    ): ClaimedEventFilter;

    "SetDefaultTokenList(address[])"(
      defaultTokenList?: null
    ): SetDefaultTokenListEventFilter;
    SetDefaultTokenList(
      defaultTokenList?: null
    ): SetDefaultTokenListEventFilter;

    "SetMerkleRoot(bytes32)"(
      merkleRoot?: BytesLike | null
    ): SetMerkleRootEventFilter;
    SetMerkleRoot(merkleRoot?: BytesLike | null): SetMerkleRootEventFilter;

    "Verified(address,address,uint256)"(
      token?: string | null,
      user?: string | null,
      amountClaimable?: null
    ): VerifiedEventFilter;
    Verified(
      token?: string | null,
      user?: string | null,
      amountClaimable?: null
    ): VerifiedEventFilter;
  };

  estimateGas: {
    claim(
      receiver: string,
      tokens: string[],
      totalAccrueds: BigNumberish[],
      proofs: BytesLike[][],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    claimVerified(
      receiver: string,
      tokens: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    claimed(
      token: string,
      user: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    merkleRoot(overrides?: CallOverrides): Promise<BigNumber>;

    verified(
      token: string,
      user: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    verify(
      user: string,
      tokens: string[],
      totalAccrueds: BigNumberish[],
      proofs: BytesLike[][],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    claim(
      receiver: string,
      tokens: string[],
      totalAccrueds: BigNumberish[],
      proofs: BytesLike[][],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    claimVerified(
      receiver: string,
      tokens: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    claimed(
      token: string,
      user: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    merkleRoot(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    verified(
      token: string,
      user: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    verify(
      user: string,
      tokens: string[],
      totalAccrueds: BigNumberish[],
      proofs: BytesLike[][],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;
  };
}
